geps/gep-91/index.md

```diff
@@ -1,22 +1,125 @@
+* Introduce an optional `subjectAltNames` field within `ClientValidationContext` that can be used to specify one or more alternate names to verify the subject identity in the certificate presented by the client. The maximum number of alternate names that can be specified is implementation defined.
+
+#### GO
+
+```go
+// ClientValidationContext holds configuration that can be used to validate the client intiating the TLS connection
+// to the Gateway.
+// By default, no client specific configuration is validated.
+type ClientValidationContext struct {
+    // CACertificateRefs contains one or more references to
+    // Kubernetes objects that contain TLS certificates of
+    // the Certificate Authorities that can be used
+    // as a trust anchor to validate the certificates presented by the client.
+    //
+    // A single CA certificate reference to a Kubernetes ConfigMap or Secret kind
+    // has "Core" support.
+    // Implementations MAY choose to support attaching multiple CA certificates to
+    // a Listener, but this behavior is implementation-specific.
+    //
+    // Support: Core - An optional single reference to a single Kubernetes Secret
+    // and to a single Kubernetes ConfigMap with the CA certificate in a key named `ca.crt`.
+    //
+    // Support: Implementation-specific (More than one reference, or other kinds
+    // of resources).
+    //
+    // References to a resource in a different namespace are invalid UNLESS there
+    // is a ReferenceGrant in the target namespace that allows the certificate
+    // to be attached. If a ReferenceGrant does not allow this reference, the
+    // "ResolvedRefs" condition MUST be set to False for this listener with the
+    // "RefNotPermitted" reason.
+    //
+    // +kubebuilder:validation:MaxItems=8
+    // +kubebuilder:validation:MinItems=1
+    CACertificateRefs []SecretObjectReference `json:”caCertificateRefs,omitempty”`
+    // SubjectAltNames contains one or more alternate names to verify
+    // the subject identity in the certificate presented by the client.
+    //
+    // Support: Core
+    //
+    // +optional
+    // +kubebuilder:validation:MinItems=1
+    SubjectAltNames []string `json:"subjectAltNames,omitempty"`
+}
+
+```
+
+#### YAML
+
+```yaml
+apiVersion: gateway.networking.k8s.io/v1beta1
+kind: Gateway
+metadata:
+  name: client-validation-basic
+spec:
+  gatewayClassName: acme-lb
+  listeners:
+  - name: foo-https
+    protocol: HTTPS
+    port: 443
+    hostname: foo.example.com
+    tls:
+      certificateRefs:
+      - kind: Secret
+        group: ""
+        name: foo-example-com-cert
+      clientValidation:
+        caCertificateRefs:
+        - kind: ConfigMap
+          group: ""
+          name: foo-example-com-ca-cert
```

Comment:
Have you considered making this a top level field instead of per listener? It's current placement requires SNI-matching prior to Client Cert validation which could be tricky. (I'm not necessarily advocating in one direction or the other here, just wanted to start a conversation).

Reply:
I'm not sure I understand, but I think all the TLS handshake info should be together like this.

Anyway, this PR is just implementing what @arkodg is proposing for GEP-91 in #2273, so design discussions should probably be there, and I'll keep this PR in sync with any changes that happen in that one.

Reply:
@robscott implementations are already doing SNI matching today to enforce listener isolation, `hostname` is defined within each `listener`, so this works fine, this even works when Gateway API listeners needs to be flattened out in the implementation data plane in case 2 listeners have the same `port`.
I didn't quite understand what the  top level definition would enable ?

Reply:
> I think all the TLS handshake info should be together like this.

Jumping forward a bit, I think it's very likely that the client cert a Gateway uses to connect to backends will be configured per Gateway (not per Listener) based on the discussion at the community meeting earlier this week. Essentially the argument is that in that case, the client cert is representing the identity of the Gateway, not any individual listener.

If we were to take that argument further, would a Gateway admin rather define the set of clients it trusts in one place or individually for each Listener? 

> implementations are already doing SNI matching today to enforce listener isolation

Some may end up with a subset of this. For example, many cloud LBs prevent domain fronting but don't expose full SNI matching config. Preventing domain fronting is actually one way to meet the requirements of SNI/hostname matching in Gateways, but that logic may not extend to things like client cert validation.

One detail that's not clear to me from the list of implementations that support client cert validation in this GEP is how many support configuring this differently per SNI on the same VIP + Port. That would help show the viability of this. For what it's worth, in GCP this config is per VIP + Port, not per SNI (https://cloud.google.com/load-balancing/docs/mtls). Of course a single implementation does not need to impact the entire design here, my hunch is that GCP may not be the only one with this limitation.

Reply:
resolving this comment since its being addressed by Rob with https://github.com/kubernetes-sigs/gateway-api/pull/2910